doc: add enabling PTM tutorial and PTM hld discussion (#6110)
projectacrn/acrn-hypervisor: doc/developer-guides/hld/hld-devicemodel.rst

   00:04.0 Ethernet controller: Red Hat, Inc. Virtio network device
   00:05.0 Serial controller: Red Hat, Inc. Virtio console

.. _PCIe PTM implementation:

> PCIE Root Port Emulation

Reviewer: Sure.  Sounds good to me.

> PCIE Root Port Emulation

Reviewer: Ubuntu kernel doesn't have these 2 config flags turned on by default and that is why we need to document these here.  In addition, we do need to make sure that IGC device driver has patch that enables PTM.  This patch is not upstreamed to mainline linux kernel yet but it is already in PKT (so it is in acrn kernel as I was told).

========================

The PCI Express (PCIe) specification defines a Precision Time Measurement (PTM)
mechanism that enables time coordination and synchronization of events across
multiple PCI components with independent local time clocks within the same
system.  Intel supports PTM on several of its systems and devices, such as PTM
root capabilities support on Whiskey Lake and Tiger Lake PCIe root ports, and
PTM device support on an Intel I225-V/I225-LM family Ethernet controller.  For
further details on PTM, please refer to the `PCIe specification
<https://pcisig.com/specifications>`_.

ACRN adds PCIe root port emulation in the hypervisor to support the PTM feature
and emulates a simple PTM hierarchy.  ACRN enables PTM in a Guest VM if the user
sets the ``enable_ptm`` option when passing through a device to a post-launched
VM. When the user enables PTM, the passthru device is connected to a virtual

> root port instead of the host bridge.  Here is an example launch script that

Reviewer: Good point.  But I suppose this has been taken care of already since launch script already has this?  Maybe revisit the document of launch script to add this.

> root port instead of the host bridge.  Here is an example launch script that

Reviewer: It sounds good to me.  But I would like to see whether @gvancuts and @wenlingz have any question after reading above paragraph?

enables PTM on a passthru Ethernet card:

.. code-block:: bash
   :emphasize-lines: 5

   acrn-dm -A -m $mem_size -s 0:0,hostbridge \
      -s 3,virtio-blk,uos-test.img \
      -s 4,virtio-net,tap0 \
      -s 5,virtio-console,@stdio:stdio_port \
      -s 6,passthru,a9/00/0,enable_ptm \
      --ovmf /usr/share/acrn/bios/OVMF.fd

> And here is the bus hierarchy from guest (as shown by the ``lspci`` command)::

Reviewer: Sounds good.

>    sudo lspci -tv

Reviewer: Sounds good.

   -[0000:00]-+-00.0  Network Appliance Corporation Device 1275
              +-03.0  Red Hat, Inc. Virtio block device
              +-04.0  Red Hat, Inc. Virtio network device
              +-05.0  Red Hat, Inc. Virtio console

>               \-06.0-[01]----00.0  Intel Corporation Device 15f2

Reviewer: I am not sure there is anything we can do (or worth to do here).  It seems SOS (or baremetal) shows the info in the same format.  For example, this is from SOS:
           +-1c.0-[a9]----00.0  Intel Corporation Device 15f2

I used Ubuntu 18.04.5.

>               \-06.0-[01]----00.0  Intel Corporation Device 15f2

Reviewer 2: OK, I think this is due to a `hwdata` package (the one that contains the `pci.ids` file) that does not yet include that device. How about we add a note stating that `Ethernet Controller I225-LM` may also appear?

PTM Implementation Notes
------------------------

To simplify PTM support implementation, the virtual root port only supports the
most basic PCIe configuration and operation, in addition to PTM capabilities.

In Guest VM post-launched scenarios, you enable PTM by setting the
``enable_ptm`` option for the pass through device (as shown above).

> .. figure:: images/PTM-hld-PTM-flow.png

Reviewer 2: Can we update the picture to use "Post-launched VM" instead of "Post Launch VM"

   :align: center
   :width: 700
   :name: ptm-flow

   PTM-enabling workflow in post-launched VM

As shown in :numref:`ptm-flow`, PTM is enabled in the root port during the
hypervisor startup. Device model then checks whether the pass through device
supports PTM requestor capabilities and whether the corresponding root port
supports PTM root capabilities, as well as some other sanity checks.  If an
error is detected during these checks, the error will be reported and ACRN will
not enable PTM on guest. This doesn’t prevent the user from launching the Guest
VM and passing through the device to the Guest VM.  If no error is detected,
device model will use ``add_vdev`` hypercall to add a virtual root port (VRP),
acting as the PTM root, to Guest before passing through the device to Guest.

.. figure:: images/PTM-hld-PTM-passthru.png
   :align: center
   :width: 700
   :name: ptm-vrp

   PTM-enabled PCI device pass-through to post-launched VM

> :numref:`ptm-vrp` show that, after enabling PTM, the passthru device connects to

Reviewer 2: ```suggestion
:numref:`ptm-vrp` shows that, after enabling PTM, the passthru device connects to
```
?

the virtual root port instead of the virtual host bridge.

To use PTM in a virtualized environment, you may want to first verify that PTM
is supported by the device and is enabled on the bare metal machine.

> If supported, follow these steps to enable PTM on post-launched guest:

Reviewer 2: ```suggestion
If supported, follow these steps to enable PTM in the post-launched guest:
```

> 1. Make sure that PTM is enabled in guest kernel.  In Linux kernel, for example,

Reviewer 2: ```suggestion
1. Make sure that PTM is enabled in the guest kernel.  In the Linux kernel, for example,
```

>    set “CONFIG_PTM=on”.

Reviewer 2: ```suggestion
   set ``CONFIG_PCIE_PTM=y``.
```

>    set ``CONFIG_PTM=y``.

Reviewer 2: ```suggestion
   set ``CONFIG_PCIE_PTM=y``.
```

> 2. Not every PCI device supports PTM.  One example that does it the Intel I225-V

Reviewer 2: ```suggestion
2. Not every PCI device supports PTM.  One example that does it is the Intel I225-V
```

>    Ethernet controller.  If you passthru this card to guest,  make sure guest

Reviewer 2: ```suggestion
   Ethernet controller.  If you passthru this card to the guest VM, make sure the guest VM
```

>    uses a version of the IGC driver that supports PTM.

Reviewer 2: What is this IGC driver? Can we include a pointer to it?

Reviewer 2: I guess IGC here refers to the Linux kernel `CONFIG_IGC` driver?

It sounds like not all versions of the IGC drivers support PTM, can we indicate which does?

> 3. In the device model launch script, add the ``enable_ptm`` option on the

Reviewer 2: ```suggestion
3. In the device model launch script, add the ``enable_ptm`` option to the
```
?

>    passthru device.  For example:

Reviewer 2: This section is copy-paste of what's already mentionned above. Do we need both?

   .. code-block:: bash
      :emphasize-lines: 5

      $ acrn-dm -A -m $mem_size -s 0:0,hostbridge \
          -s 3,virtio-blk,uos-test.img \
          -s 4,virtio-net,tap0 \
          -s 5,virtio-console,@stdio:stdio_port \
          -s 6,passthru,a9/00/0,enable_ptm \
          --ovmf /usr/share/acrn/bios/OVMF.fd \

4. You can check that PTM is correctly enabled on guest by displaying the PCI
   bus hiearchy on the guest using the ``lspci`` command:

   .. code-block:: bash
      :emphasize-lines: 12,20

>       $ sudo lspci -tv

Reviewer 2: ```suggestion
      lspci -tv
```

        -[0000:00]-+-00.0  Network Appliance Corporation Device 1275
         +-03.0  Red Hat, Inc. Virtio block device
         +-04.0  Red Hat, Inc. Virtio network device
         +-05.0  Red Hat, Inc. Virtio console

>          \-06.0-[01]----00.0  Intel Corporation Device 15f2

Reviewer 3: I am confused here, as "-s 6,passthru,a9/00/0,enable_ptm \" in launch script, why 00:06.0 is assigned to vrp in guest rather than to the passthru device? It seems the passthru device is always assigned with bdf 01:00.0, which is hardcoded in acrn. 
Why not dynamically assign bdf for passthru device decided by dm option "-s <slot>" (aka. by user) and hardcode bdf for vrp?

Reviewer: In order to enable PTM for guest, we need to add a virtual root port, and this virtual root port uses the passthru device's bdf (it doesn't have to but I think it is better than a random assigned bdf).  And now passthru device is connected to virtual root (a different bus) instead of host bridge and that is why the passthru device gets 01:00.0 (on bus 1 now, instead of bus 0).  Why it is hardcoded, mostly because this is what we can support in the timeframe of release 2.5, i.e., just 1 pass thru device with PTM.  In the future, for example, when there are more PTM passthru devices, we need to change this (at least the "bus" part of device's bdf).

>       $ sudo lspci -vv # (Only relevant output is shown)

Reviewer 2: ```suggestion
      sudo lspci -vv # (Only relevant output is shown)
```
